File: mac/Sources/LegionControlCore/WakePlan.swift
import Foundation

/// How to wake one machine from where this device happens to be.
///
/// A magic packet is a broadcast on a local network, so it only works from that network. One helper
/// in one house cannot wake a tower in another, whatever a config says. The plan is therefore
/// ordered: send it directly if this device is on the right network, otherwise ask a machine that is
/// — in the order the setup lists them, until one of them does it.
///
/// Two rules the plan never breaks. Nothing is ever woken in order to serve as a helper: a wake
/// cascade would turn an always-off tower into an always-on one by accident, which is the opposite
/// of what the user asked for. And when nothing can be done, the app says so with a reason for each
/// helper rather than pretending the packet went somewhere.
enum WakePlan {

    /// One thing to try, in order.
    enum Step: Sendable, Equatable {
        /// Send the packet from this device.
        case direct(broadcasts: [String], ports: [UInt16])
        /// Ask a machine to send it, by running a configured action there.
        case helper(WakeHelper, machineName: String)
    }

    /// Why a step is not available.
    struct Unavailable: Sendable, Equatable, Identifiable {
        var what: String
        var reason: String
        /// The machine that would have to be woken first, when that is the obstacle. Never acted on
        /// automatically: it is offered as a separate, explicit step.
        var wakeableFirst: String?

        var id: String { what }
    }

    struct Plan: Sendable, Equatable {
        var steps: [Step]
        var unavailable: [Unavailable]
        /// Where this device thinks it is, for the explanation line.
        var placement: SiteAwareness.Placement

        var isEmpty: Bool { steps.isEmpty }

        /// The sentence shown when there is nothing to try.
        var nothingToTry: String {
            guard steps.isEmpty else { return "" }
            if unavailable.isEmpty {
                return "There is no way to wake this machine from here: nothing on its network is configured to send the packet."
            }
            return "Nothing here can wake it right now. " + unavailable.map { "\($0.what): \($0.reason)" }
                .joined(separator: " ")
        }
    }

    /// What a helper machine looked like at the last reading, so a plan can skip one that is known
    /// to be off rather than spending a round trip finding out.
    struct HelperState: Sendable, Equatable {
        var isAwake: Bool
        var isCommandable: Bool
        var canBeWoken: Bool
        var lastCheckedAt: Date?
    }

    /// Build the plan.
    static func plan(
        for machine: Machine,
        in config: ControllerConfig,
        placement: SiteAwareness.Placement,
        addresses: [String],
        helperStates: [String: HelperState]
    ) -> Plan {
        guard let wake = machine.wake else {
            return Plan(steps: [], unavailable: [], placement: placement)
        }

        var steps: [Step] = []
        var unavailable: [Unavailable] = []

        // 1. Directly, when this device is on the machine's network. The site's own broadcast is the
        //    fallback for a machine that names none, which is the usual case once sites exist.
        let site = machine.site.flatMap { config.site(id: $0) }
        let onSite: Bool = {
            if let machineSite = machine.site { return placement.effective == machineSite }
            return SiteAwareness.matchesLANPrefix(wake.lanPrefix, addresses: addresses)
        }()

        if onSite {
            var broadcasts = wake.broadcast
            if broadcasts == ["255.255.255.255"], let site, !site.broadcast.isEmpty {
                broadcasts = site.broadcast
            }
            steps.append(.direct(broadcasts: broadcasts, ports: wake.ports))
        } else if placement.isAmbiguous {
            unavailable.append(Unavailable(
                what: "Sending it from here",
                reason: "this device's addresses match more than one of the configured sites, so a broadcast could go to the wrong network. Confirm which site this is to send it directly.",
                wakeableFirst: nil
            ))
        } else {
            let where_ = site?.name ?? machine.name
            unavailable.append(Unavailable(
                what: "Sending it from here",
                reason: "this device is not on \(where_)'s network, and a magic packet does not cross networks.",
                wakeableFirst: nil
            ))
        }

        // 2. Helpers, in the order the setup lists them.
        for helper in wake.orderedHelpers {
            guard let host = config.machine(id: helper.machine) else {
                unavailable.append(Unavailable(
                    what: helper.machine,
                    reason: "the setup lists it as a helper but has no such machine.",
                    wakeableFirst: nil
                ))
                continue
            }
            let state = helperStates[helper.machine]
            if let state, state.lastCheckedAt != nil, !state.isAwake {
                unavailable.append(Unavailable(
                    what: host.name,
                    reason: host.alwaysOn == true
                        ? "it is marked as always on but did not answer at the last reading."
                        : "it is asleep or off.",
                    // Offered, never taken. Waking a helper to wake something else is a cascade, and
                    // a cascade is how the machine you were trying not to run ends up running.
                    wakeableFirst: host.wake != nil ? host.name : nil
                ))
                continue
            }
            if let state, state.lastCheckedAt != nil, state.isAwake, !state.isCommandable {
                unavailable.append(Unavailable(
                    what: host.name,
                    reason: "it is awake but its control agent is not answering.",
                    wakeableFirst: nil
                ))
                continue
            }
            steps.append(.helper(helper, machineName: host.name))
        }

        if steps.isEmpty, wake.orderedHelpers.isEmpty, !onSite {
            unavailable.append(Unavailable(
                what: "A helper",
                reason: "no machine on \(site?.name ?? "that network") is configured to send the packet.",
                wakeableFirst: nil
            ))
        }

        return Plan(steps: steps, unavailable: unavailable, placement: placement)
    }

    /// Whether a helper failure is safe to move past.
    ///
    /// A `wol` action is a declared, idempotent UDP send: repeating it costs three more datagrams
    /// and nothing else, so an ambiguous failure may be followed by the next helper. Any other
    /// configured action may do anything at all, and an ambiguous outcome there has to be settled —
    /// by reading its operation record when the helper is reachable again — before something else
    /// is tried. Otherwise "the link dropped" turns one requested action into two performed ones.
    static func mayTryNextHelper(after failure: AgentFailure, actionIsDeclaredWakePacket: Bool) -> Bool {
        switch failure.dispatch {
        case .never:
            return true
        case .unknown, .acknowledged:
            return actionIsDeclaredWakePacket
        }
    }
}

File: android/app/src/main/java/com/x1f4r/legioncontrol/agent/WakePlan.kt
package com.x1f4r.legioncontrol.agent

import com.x1f4r.legioncontrol.data.SiteConfig
import com.x1f4r.legioncontrol.net.SiteMatch

/**
 * How to wake one machine from where this device is standing.
 *
 * There is no longer one house and no longer one machine that is always on. A magic packet is a
 * link-local broadcast, so waking a machine in another building is not a matter of trying harder: it
 * needs somebody already on that network to send it. That somebody is a helper, helpers can themselves be asleep,
 * and so there is an order and there are failures worth explaining rather than retrying.
 *
 * Nothing here wakes a helper. The user asked not to be pushed towards leaving machines on, and a
 * cascade that wakes the tower to wake the laptop turns an always-off tower into an always-on one by
 * accident. Where a helper is itself wakeable, the app offers that as a separate thing to press.
 */
sealed interface WakePlan {
    /** This device is on the machine's own network. Send the packet from here. */
    data class Direct(
        val broadcasts: List<String>,
        val ports: List<Int>,
        /** The prefixes the socket has to leave by, so a tunnel does not swallow the broadcast. */
        val prefixes: List<String>,
        val why: String,
    ) : WakePlan

    /** Ask each of these in turn, stopping at the first that says it sent something. */
    data class ViaHelpers(val helpers: List<WakeHelper>, val why: String) : WakePlan

    /**
     * Nothing here can wake it, and this says why for each thing that was considered.
     *
     * [wakeableHelpers] are helpers that could themselves be woken, offered as a separate explicit
     * action rather than done automatically.
     */
    data class Impossible(
        val reasons: List<String>,
        val wakeableHelpers: List<String>,
    ) : WakePlan

    /** The machine has no wake configuration at all, so the action is not drawn. */
    data object NotConfigured : WakePlan
}

/**
 * Works out the plan.
 *
 * [siteMatch] is deliberately allowed to be uncertain. An ambiguous site means several configured
 * sites share a private range, and the safe reading of that is "not on site": sending a broadcast at
 * the wrong house achieves nothing, whereas asking a helper that is genuinely on the target's LAN
 * works from anywhere. So ambiguity falls through to helpers rather than guessing.
 */
fun planWake(
    target: Machine,
    sites: List<SiteConfig>,
    siteMatch: SiteMatch,
    /** Every machine in the document, so a helper can be named and its own wake seen. */
    machines: List<Machine>,
    /** The prefixes this device holds an address in, for the unplaced single-house case. */
    localAddresses: List<String>,
): WakePlan {
    val wake = target.wake ?: return WakePlan.NotConfigured
    val site = target.siteId?.let { id -> sites.firstOrNull { it.id == id } }

    val onTargetNetwork = when {
        // The machine names a site, and this device is on it.
        site != null -> siteMatch.site?.id == site.id
        // No sites in the document: the old single-house arrangement, one prefix per machine.
        !wake.lanPrefix.isNullOrBlank() -> localAddresses.any { it.startsWith(wake.lanPrefix) }
        else -> false
    }

    val broadcasts = wake.broadcasts.ifEmpty { site?.broadcast.orEmpty() }
    val prefixes = site?.lanPrefixes ?: listOfNotNull(wake.lanPrefix)

    if (onTargetNetwork && broadcasts.isNotEmpty()) {
        return WakePlan.Direct(
            broadcasts = broadcasts,
            ports = wake.ports,
            prefixes = prefixes,
            why = site?.let { "This device is on ${it.displayName}." }
                ?: "This device is on ${target.name}'s network.",
        )
    }

    if (wake.helpers.isNotEmpty()) {
        val why = when {
            onTargetNetwork ->
                "This device is on the right network and ${target.name} has no broadcast address " +
                    "configured, so a helper is asked instead."

            siteMatch is SiteMatch.Ambiguous ->
                "This device's address matches more than one site, so it cannot be trusted to be on " +
                    "${target.name}'s network. A helper that is already there is asked instead."

            else -> "This device is not on ${target.name}'s network, so a helper that is asks for it."
        }
        return WakePlan.ViaHelpers(wake.helpers, why)
    }

    val reasons = buildList {
        if (onTargetNetwork && broadcasts.isEmpty()) {
            add("${target.name} has no broadcast address configured, so a packet has nowhere to go.")
        } else {
            add(
                when (siteMatch) {
                    is SiteMatch.Ambiguous -> siteMatch.explain()
                    else -> "This device is not on ${target.name}'s network, and a magic packet is a " +
                        "LAN broadcast that a tunnel cannot carry."
                },
            )
        }
        add("No wake helper is configured for ${target.name}.")
    }
    return WakePlan.Impossible(reasons, wakeableHelpers = emptyList())
}

/**
 * What to say and offer when every helper has been tried and none of them worked.
 *
 * [attempts] is one line per helper, in the order they were asked. The offer to wake a helper first
 * is deliberately a separate press: it is a second machine being started, which is exactly the
 * energy cost the user asked not to be signed up to silently.
 */
fun helperFailure(
    target: Machine,
    attempts: List<String>,
    machines: List<Machine>,
    helpers: List<WakeHelper>,
): WakePlan.Impossible {
    val wakeable = helpers.mapNotNull { helper ->
        machines.firstOrNull { it.id == helper.machineId && it.wake != null }?.name
    }.distinct()
    return WakePlan.Impossible(
        reasons = listOf("Nothing could wake ${target.name} from here.") + attempts,
        wakeableHelpers = wakeable,
    )
}
